src/libcrun/handlers/wasmtime.c

```diff
@@ -0,0 +1,241 @@
+  wasmtime_module_new = dlsym (cookie, "wasmtime_module_new");
+  wasi_config_inherit_argv = dlsym (cookie, "wasi_config_inherit_argv");
+  wasi_config_inherit_stdout = dlsym (cookie, "wasi_config_inherit_stdout");
+  wasi_config_inherit_stdin = dlsym (cookie, "wasi_config_inherit_stdin");
+  wasi_config_inherit_stderr = dlsym (cookie, "wasi_config_inherit_stderr");
+  wasi_config_inherit_env = dlsym (cookie, "wasi_config_inherit_env");
+  wasmtime_context_set_wasi = dlsym (cookie, "wasmtime_context_set_wasi");
+  wasmtime_linker_module = dlsym (cookie, "wasmtime_linker_module");
+  wasmtime_linker_get_default = dlsym (cookie, "wasmtime_linker_get_default");
+  wasmtime_func_call = dlsym (cookie, "wasmtime_func_call");
+  wasmtime_module_delete = dlsym (cookie, "wasmtime_module_delete");
+  wasmtime_store_delete = dlsym (cookie, "wasmtime_store_delete");
+
+  if (wasm_engine_new == NULL || wasm_engine_delete == NULL || wasm_byte_vec_delete == NULL
+      || wasm_byte_vec_new_uninitialized == NULL || wasi_config_new == NULL || wasmtime_store_new == NULL
+      || wasmtime_store_context == NULL || wasmtime_linker_new == NULL || wasmtime_linker_define_wasi == NULL
+      || wasmtime_module_new == NULL || wasi_config_inherit_argv == NULL || wasi_config_inherit_stdout == NULL
+      || wasi_config_inherit_stdin == NULL || wasi_config_inherit_stderr == NULL
+      || wasi_config_inherit_env == NULL || wasmtime_context_set_wasi == NULL
+      || wasmtime_linker_module == NULL || wasmtime_linker_get_default == NULL || wasmtime_func_call == NULL
+      || wasmtime_module_delete == NULL || wasmtime_store_delete == NULL)
+    error (EXIT_FAILURE, 0, "could not find symbol in `libwasmtime.so`");
+
+  // Set up wasmtime context
+  wasm_engine_t *engine = wasm_engine_new ();
+  assert (engine != NULL);
+  wasmtime_store_t *store = wasmtime_store_new (engine, NULL, NULL);
+  assert (store != NULL);
+  wasmtime_context_t *context = wasmtime_store_context (store);
+
+  // Link with wasi functions defined
+  wasmtime_linker_t *linker = wasmtime_linker_new (engine);
+  wasmtime_error_t *err = wasmtime_linker_define_wasi (linker);
+  if (err != NULL)
+    error (EXIT_FAILURE, -1, "failed to link wasi");
+
+  wasm_byte_vec_t wasm;
+  // Load and parse container entrypoint
+  FILE *file = fopen (pathname, "rb");
+  if (! file)
+    {
+      error (EXIT_FAILURE, -1, "error loading entrypoint");
+    }
+  fseek (file, 0L, SEEK_END);
+  size_t file_size = ftell (file);
+  wasm_byte_vec_new_uninitialized (&wasm, file_size);
+  fseek (file, 0L, SEEK_SET);
+  if (fread (wasm.data, file_size, 1, file) != 1)
+    {
+      error (EXIT_FAILURE, -1, "error load");
+    }
+  fclose (file);
+
+  // Compile wasm modules
+  wasmtime_module_t *module = NULL;
+  err = wasmtime_module_new (engine, (uint8_t *) wasm.data, wasm.size, &module);
+  if (! module)
+    error (EXIT_FAILURE, -1, "failed to compile module");
```

Comment:
could we give a better error if we consider the value of `err`?

-1 is not a valid errno, you need to specify 0 if there is no errno value.

The same comment about `-1` and `err` holds for other instances